File: copasi/bindings/python/examples/run_mca.py
#!/usr/bin/env python

# This example loads an SBML file and runs the MCA task on it,
# finally the result is plotted.
#


from __future__ import print_function
import sys
from COPASI import *
from types import *

dataModel = None


def print_annotated_matrix(title, annotated_matrix):
    # type: (str, CArrayAnnotation) -> None
    """ Utility function printing an annotated matrix"""

    #print(annotated_matrix.printToString())
    #return
    
    print(title)
    print('==========')

    size = annotated_matrix.size()
    if len(size) != 2:
        print ("  This simple function only deals with two dimensional matrices")
        return

    rows = size[0]
    columns = size[1]
    print ("Size of the matrix is: {0} rows x {1} columns".format(rows, columns))

    row_headers = annotated_matrix.getAnnotationsString(0)
    col_headers = annotated_matrix.getAnnotationsString(1)

    #print column headers
    print ("\t\t", end='')
    for i in range(columns):
        print("{0}\t".format(col_headers[i]), end='')
    print()

    for j in range(rows):
        for i in range(columns):
            if i == 0:
                print("{0}\t".format(row_headers[j]), end='')
            current = annotated_matrix.get(j,i)
            print("{0}\t".format(current), end='')
        print()
    print()
    pass


def run_mca(file_name):
    # type: (str) -> None
    """ This function runs the MCA task on the given SBML file"""

    global dataModel
    if dataModel is None:
        dataModel = CCopasiRootContainer.addDatamodel()
        assert (isinstance(dataModel, CCopasiDataModel))

    # load COPASI file
    try: 
      if file_name.endswith('.cps'):
          if not dataModel.loadModel(file_name):
              print("Could not load COPASI file due to:")
              print(CCopasiMessage.getAllMessageText())
              sys.exit(1)
      # load sbml file
      elif not dataModel.importSBML(file_name):
          print("Could not load SBML file due to:")
          print(CCopasiMessage.getAllMessageText())
          sys.exit(1)
    except:
      print("Could not load file due to:")
      print(CCopasiMessage.getAllMessageText())
      sys.exit(1)

    # setup mca task
    task = dataModel.getTask("Metabolic Control Analysis")
    assert (isinstance(task, CMCATask))
    # mark task as executable
    task.setScheduled(True)
    problem = task.getProblem()
    assert (isinstance(problem, CMCAProblem))
    # specify that we want to perform steady state analysis
    problem.setSteadyStateRequested(True)

    # run mca task
    if not task.initialize(CCopasiTask.OUTPUT_UI):
        print ("could not initialize mca task")
        print(CCopasiMessage.getAllMessageText())
        sys.exit(2)

    if not task.processWithOutputFlags(True, CCopasiTask.OUTPUT_UI):
        print ("could not run mca task")
        print (task.getProcessError())
        sys.exit(3)

    # print results
    method = task.getMethod()
    assert (isinstance(method, CMCAMethod))
    print_annotated_matrix("Scaled Concentration Control Coefficients", method.getScaledConcentrationCCAnn())
    print_annotated_matrix("Scaled Flux Control Coefficients", method.getScaledFluxCCAnn())
    print_annotated_matrix("Scaled Elasticities", method.getScaledElasticitiesAnn())

    pass


if __name__ == "__main__":
    if len(sys.argv) != 2:
        print("usage: run_mca <sbml file | copasi file>")
        sys.exit(1)
    else:
        run_mca(sys.argv[1])

File: copasi/bindings/csharp/examples/run_mca.cs
using org.COPASI;
using System;
using System.Diagnostics;

public class Run_MCA
{

  static CCopasiDataModel dataModel = null;

  /// <summary>
  /// Utility function printing an annotated matrix
  /// </summary>
  /// <param name="title"></param>
  /// <param name="annotated_matrix"></param>
  public static void print_annotated_matrix(string title, CArrayAnnotation annotated_matrix)
  {

    Console.WriteLine(title);
    Console.WriteLine("==========");

    var size = annotated_matrix.size();
    if (size.Count != 2)
    {
      Console.WriteLine("  This simple function only deals with two dimensional matrices");
      return;
    }

    var rows = size[0];
    var columns = size[1];

    Console.WriteLine(string.Format("Size of the matrix is: {0} rows x {1} columns", rows, columns));

    var row_headers = annotated_matrix.getAnnotationsString(0);
    var col_headers = annotated_matrix.getAnnotationsString(1);

    // print column headers
    Console.Write("\t\t");
    for (int i = 0; i < columns; ++i)
      Console.Write(string.Format("{0}\t", col_headers[i]));
    Console.WriteLine();

    for (int j = 0; j < rows; ++j)
    {


      for (int i = 0; i < columns; ++i)
      {
        if (i == 0)
          Console.Write(string.Format("{0}\t", row_headers[j]));

        var current_object = annotated_matrix.getObject(new CCopasiObjectName(string.Format("[{0}][{1}]", j, i)));
        var current = current_object.printToString();
        Console.Write(string.Format("{0}\t", current));
      }
      Console.WriteLine();
    }
    Console.WriteLine();
  }

  /// <summary>
  /// This function runs the MCA task on the given SBML file
  /// </summary>
  /// <param name="file_name"></param>
  public static void run_mca(string file_name)
  {
    if (dataModel == null)
      dataModel = CCopasiRootContainer.addDatamodel();

    // load COPASI file
    if (file_name.EndsWith(".cps"))
    {
      if (!dataModel.loadModel(file_name))
      {
        Console.WriteLine("Could not load COPASI file due to:");
        Console.WriteLine(CCopasiMessage.getAllMessageText());
        Environment.Exit(1);
      }
    }
    // load sbml file
    else if (!dataModel.importSBML(file_name))
    {
      Console.WriteLine("Could not load SBML file due to:");
      Console.WriteLine(CCopasiMessage.getAllMessageText());
      Environment.Exit(1);
    }

    // setup mca task
    var task = dataModel.getTask("Metabolic Control Analysis") as CMCATask;
    //# mark task as executable
    task.setScheduled(true);
    var problem = task.getProblem() as CMCAProblem;
    // specify that we want to perform steady state analysis
    problem.setSteadyStateRequested(true);

    // run mca task
    if (!task.initialize((int)CCopasiTask.OUTPUT_UI))
    {
      Console.WriteLine("could not initialize mca task");
      Console.WriteLine(CCopasiMessage.getAllMessageText());
      Environment.Exit(2);
    }
    if (!task.processWithOutputFlags(true, (int)CCopasiTask.OUTPUT_UI))
    {
      Console.WriteLine("could not run mca task");
      Console.WriteLine(task.getProcessError());
      Environment.Exit(3);
    }
    // messages
    Console.WriteLine(CCopasiMessage.getAllMessageText());
    Console.WriteLine(task.getProcessError());

    // print results
    var method = task.getMethod() as CMCAMethod;
    print_annotated_matrix("Scaled Concentration Control Coefficients", method.getScaledConcentrationCCAnn());
    print_annotated_matrix("Scaled Flux Control Coefficients", method.getScaledFluxCCAnn());
    print_annotated_matrix("Scaled Elasticities", method.getScaledElasticitiesAnn());

  }

  public static int Main(String[] args)
  {
    if (args.Length != 1)
    {
      Console.WriteLine("usage: run_mca <sbml file | copasi file>");
      return 1;
    }
    run_mca(args[0]);
    return 0;
  }
}